### ISSUE TITLE
Developer's logbook

### ISSUE THREAD

**Issue by mmascher** [dmwm/CRABServer#4583]
In this github issue I would like to collect useful information for developers who are running their own version of CRAB3 (from the client to the REST/Server/Postjob etc)

**Comment by mmascher** [comment]
You will need to update the database with: https://github.com/dmwm/CRABServer/blob/master/etc/updateOracle.sql#L40 if you use the master branch

**Comment by mmascher** [comment]
Hi,

as of this morning oracle database update you will need to update the tnsnames.ora in your REST instances.

You can take the one on lxplus in /etc and replace the one used by the VM. that's what I did:

```
sudp scp mmascher@lxplus:/etc/tnsnames.ora srv/HG1410d/sw.pre.mmascher/slc6_amd64_gcc481/cms/oracle-env/29/etc/tnsnames.ora
```

**Comment by mmascher** [comment]
This afternoon I merged a pull from justas that requires `ALTER TABLE tasks ADD (tm_schedd VARCHAR(255));`

Also, if you update the code remember to update and restart both REST and TW

**Comment by mmascher** [comment]
Hey @nizamyusli , I had the same issue you had yesterday, when you run an updated version of the REST and old version of the TW (*). 

I wanted to tell you that I had to clean my database to solve it (I think I executed `delete from tasks`)

(*)
```
The cachefile /tmp/mmascher/.wmcore_cache_N74Adr/cmsweb.cern.ch/-5861594372472728703_GET_lfn2pfn does not exist and the service at https://cmsweb.cern.ch/phedex/datasvc/json/prod/lfn2pfn is unavailable - it returned 400 because Bad Request
 with result: The 'lfn' parameter ("/store/user/mmascher/DoubleMuParked/hi/150214_181528:mmascher/0000") to (eval) did not pass the 'lfn' callback\n\n


Traceback (most recent call last):
  File "/afs/cern.ch/user/m/mmascher/repos/CRABServer/src/python/TaskWorker/Actions/DagmanCreator.py", line 328, in resolvePFNs
    pfn_info = self.phedex.getPFN(nodes=dest_sites_, lfns=lfns)
  File "/afs/cern.ch/user/m/mmascher/repos/WMCore/src/python/WMCore/Services/PhEDEx/PhEDEx.py", line 389, in getPFN
    data = self._getResult('lfn2pfn', args = input_dict, verb = 'GET')
  File "/afs/cern.ch/user/m/mmascher/repos/WMCore/src/python/WMCore/Services/PhEDEx/PhEDEx.py", line 52, in _getResult
    f = self.refreshCache(file, callname, args, verb = verb)
  File "/afs/cern.ch/user/m/mmascher/repos/WMCore/src/python/WMCore/Services/Service.py", line 210, in refreshCache
    self.getData(cachefile, url, inputdata, incoming_headers, encoder, decoder, verb, contentType)
  File "/afs/cern.ch/user/m/mmascher/repos/WMCore/src/python/WMCore/Services/Service.py", line 314, in getData
    raise he
HTTPException
```

**Comment by nizamyusli** [comment]
Noted. Thanks Marco.

**Comment by annawoodard** [comment]
I think we need to update tnsnames.ora again:
```
sudo scp awoodard@lxplus:/etc/tnsnames.ora /data/srv/HG1502b/sw.pre/slc6_amd64_gcc481/cms/oracle-env/29/etc/
```

**Comment by juztas** [comment]
Please all take a look into this: https://hypernews.cern.ch/HyperNews/CMS/get/crabDevelopment/2307.html

**Comment by mmascher** [comment]
This pull https://github.com/dmwm/CRABServer/pull/4833 once merged will require a new version of WMCore, otherwise you will not find the `getAllPhEDExNodeNames` method.

I am waiting for merging this.

**Comment by mmascher** [comment]
Merged https://github.com/dmwm/CRABServer/pull/4833

Things will break if you use the master branch now, I will validate the new WMCore tag tomorrow.

**Comment by mmascher** [comment]
Dear all,

I have created 3.3.1507.rc1 and I uploaded them on comp.pre.mmascher

Notice that if you use the master branch for development and you take the code from git then you will need the new WMCore (I am using tag `1.0.8.pre6`), and the new WMCore requires an updated dbsclient (I suggest you to reinstall the taskworker)

**Comment by belforte** [comment]
will TW install from scratch pull new WMCore ?

**Comment by belforte** [comment]
I installed this TW tag on my new VM following
https://twiki.cern.ch/twiki/bin/view/CMSPublic/CMSCrabTaskWorker

It did not pull WMCore [1]
Is that because I have already the new WMCore from the previous
CrabServer install on same VM ?

I have
/data/srv/current/sw.pre/slc6_amd64_gcc481/cms/crabserver/3.3.1506.rc2-comp/lib/python2.6/site-packages/WMCore
But do not know how to tell which version of WMCore it is.

stefano

[1]
belforte@stefanovm/data> apt-get -y install cms+crabtaskworker+$RELEASE
Reading Package Lists... Done
Building Dependency Tree... Done
The following extra packages will be installed:
   cms+dbs-client+DBS_2_1_9-comp8 cms+dbs3-client+3.2.11d-comp2
   cms+dbs3-pycurl-client+3.2.11d-comp2 cms+dls-client+DLS_1_1_3-comp8
   external+boost+1.57.0-comp5 external+bz2lib+1.0.6 external+c-ares+1.10.0
   external+cherrypy+3.2.2-comp8 external+condor+8.3.1-comp6 external+curl+7.35.0-comp6
   external+db4+4.8.30 external+expat+2.0.1-comp external+gcc+4.8.1 external+gdbm+1.8.3
   external+libtool+2.4.2 external+libuuid+2.22.2 external+libxml2+2.9.1
   external+openssl+1.0.1o external+p5-archive-tar+1.76-comp
   external+p5-extutils-makemaker+6.63_02-comp external+p5-io-zlib+1.10-comp
   external+p5-package-constants+0.02-comp external+p5-time-hires+1.9715-comp
   external+pcre+7.9-comp external+py2-cjson+1.0.5-comp8 external+py2-httplib2+0.7.1-comp8
   external+py2-pycurl+7.19.3-comp8 external+py2-pyxml+0.8.4-comp8 external+py2-zsi+2.0-comp8
   external+python+2.6.8-comp8 external+sqlite+3.8.1 external+zlib+1.2.8
The following NEW packages will be installed:
   cms+crabtaskworker+3.3.1507.rc1 cms+dbs-client+DBS_2_1_9-comp8
   cms+dbs3-client+3.2.11d-comp2 cms+dbs3-pycurl-client+3.2.11d-comp2
   cms+dls-client+DLS_1_1_3-comp8 external+boost+1.57.0-comp5 external+bz2lib+1.0.6
   external+c-ares+1.10.0 external+cherrypy+3.2.2-comp8 external+condor+8.3.1-comp6
   external+curl+7.35.0-comp6 external+db4+4.8.30 external+expat+2.0.1-comp external+gcc+4.8.1
   external+gdbm+1.8.3 external+libtool+2.4.2 external+libuuid+2.22.2 external+libxml2+2.9.1
   external+openssl+1.0.1o external+p5-archive-tar+1.76-comp
   external+p5-extutils-makemaker+6.63_02-comp external+p5-io-zlib+1.10-comp
   external+p5-package-constants+0.02-comp external+p5-time-hires+1.9715-comp
   external+pcre+7.9-comp external+py2-cjson+1.0.5-comp8 external+py2-httplib2+0.7.1-comp8
   external+py2-pycurl+7.19.3-comp8 external+py2-pyxml+0.8.4-comp8 external+py2-zsi+2.0-comp8
   external+python+2.6.8-comp8 external+sqlite+3.8.1 external+zlib+1.2.8
0 upgraded, 33 newly installed, 0 removed and 0 not upgraded.
Need to get 188MB of archives.
After unpacking 744MB of additional disk space will be used.
Get:1 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+gcc+4.8.1 1-1 [136MB]
Get:2 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+zlib+1.2.8 1-1 [86.4kB]
Get:3 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+bz2lib+1.0.6 1-1 [383kB]
Get:4 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+openssl+1.0.1o 1-1 [1602kB]
Get:5 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+expat+2.0.1-comp 1-1 [217kB]
Get:6 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+db4+4.8.30 1-1 [838kB]
Get:7 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+gdbm+1.8.3 1-1 [18.4kB]
Get:8 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+sqlite+3.8.1 1-1 [550kB]
Get:9 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+python+2.6.8-comp8 1-1 [13.4MB]
Get:10 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+c-ares+1.10.0 1-1 [47.5kB]
Get:11 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+p5-extutils-makemaker+6.63_02-comp 1-1 [270kB]
Get:12 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+curl+7.35.0-comp6 1-1 [329kB]
Get:13 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+py2-pyxml+0.8.4-comp8 1-1 [1023kB]
Get:14 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+p5-io-zlib+1.10-comp 1-1 [12.3kB]
Get:15 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+p5-package-constants+0.02-comp 1-1 [8511B]
Get:16 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+py2-cjson+1.0.5-comp8 1-1 [28.8kB]
Get:17 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+py2-zsi+2.0-comp8 1-1 [385kB]
Get:18 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+py2-pycurl+7.19.3-comp8 1-1 [118kB]
Get:19 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+libtool+2.4.2 1-1 [492kB]
Get:20 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+libuuid+2.22.2 1-1 [86.7kB]
Get:21 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+libxml2+2.9.1 1-1 [755kB]
Get:22 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+p5-time-hires+1.9715-comp 1-1 [64.3kB]
Get:23 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+pcre+7.9-comp 1-1 [151kB]
Get:24 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+p5-archive-tar+1.76-comp 1-1 [38.9kB]
Get:25 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+boost+1.57.0-comp5 1-1 [11.2MB]
Get:26 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/cms 
cms+dbs-client+DBS_2_1_9-comp8 1-1 [638kB]
Get:27 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/cms 
cms+dbs3-pycurl-client+3.2.11d-comp2 1-1 [20.2kB]
Get:28 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+cherrypy+3.2.2-comp8 1-1 [718kB]
Get:29 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+py2-httplib2+0.7.1-comp8 1-1 [71.3kB]
Get:30 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/cms 
cms+dls-client+DLS_1_1_3-comp8 1-1 [133kB]
Get:31 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/cms 
cms+dbs3-client+3.2.11d-comp2 1-1 [50.4kB]
Get:32 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/external 
external+condor+8.3.1-comp6 1-1 [7377kB]
Get:33 http://cmsrep.cern.ch cmssw/comp.pre.mmascher/apt/slc6_amd64_gcc481/cms 
cms+crabtaskworker+3.3.1507.rc1 1-1 [11.1MB]
Fetched 188MB in 20s (9006kB/s)
Executing RPM 
(/data/srv/TaskManager/3.3.1507.rc1/slc6_amd64_gcc481/external/apt/0.5.16-comp2/bin/rpm-wrapper -Uvh 
-r /data/srv/TaskManager/3.3.1507.rc1 --force --prefix /data/srv/TaskManager/3.3.1507.rc1 --ignoreos 
--ignorearch --force --prefix /data/srv/TaskManager/3.3.1507.rc1 --ignoreos --ignorearch 
--oldpackage)...
Preparing...                ########################################### [100%]
    1:external+expat+2.0.1-co########################################### [  3%]
    2:external+gcc+4.8.1     ########################################### [  6%]
    3:external+zlib+1.2.8    ########################################### [  9%]
    4:external+bz2lib+1.0.6  ########################################### [ 12%]
    5:external+openssl+1.0.1o########################################### [ 15%]
    6:external+db4+4.8.30    ########################################### [ 18%]
    7:external+gdbm+1.8.3    ########################################### [ 21%]
    8:external+sqlite+3.8.1  ########################################### [ 24%]
    9:external+python+2.6.8-c########################################### [ 27%]
   10:external+c-ares+1.10.0 ########################################### [ 30%]
   11:external+p5-extutils-ma########################################### [ 33%]
   12:external+curl+7.35.0-co########################################### [ 36%]
   13:external+py2-pyxml+0.8.########################################### [ 39%]
   14:external+py2-zsi+2.0-co########################################### [ 42%]
   15:external+py2-pycurl+7.1########################################### [ 45%]
   16:external+p5-io-zlib+1.1########################################### [ 48%]
   17:external+p5-package-con########################################### [ 52%]
   18:external+py2-cjson+1.0.########################################### [ 55%]
   19:cms+dbs3-pycurl-client+########################################### [ 58%]
   20:external+p5-archive-tar########################################### [ 61%]
   21:cms+dbs-client+DBS_2_1_########################################### [ 64%]
   22:external+p5-time-hires+########################################### [ 67%]
   23:external+boost+1.57.0-c########################################### [ 70%]
   24:external+pcre+7.9-comp ########################################### [ 73%]
   25:external+libxml2+2.9.1 ########################################### [ 76%]
   26:external+libtool+2.4.2 ########################################### [ 79%]
   27:external+libuuid+2.22.2########################################### [ 82%]
   28:external+condor+8.3.1-c########################################### [ 85%]
   29:cms+dls-client+DLS_1_1_########################################### [ 88%]
   30:cms+dbs3-client+3.2.11d########################################### [ 91%]
   31:external+cherrypy+3.2.2########################################### [ 94%]
   32:external+py2-httplib2+0########################################### [ 97%]
   33:cms+crabtaskworker+3.3.########################################### [100%]
Done.
belforte@stefanovm/data>

On 06/29/2015 09:48 AM, Marco Mascheroni wrote:
> Dear all,
>
> I have created 3.3.1507.rc1 and I uploaded them on comp.pre.mmascher
>
> Notice that if you use the master branch for development and you take the code from git then you
> will need the new WMCore (I am using tag |1.0.8.pre6|), and the new WMCore requires an updated
> dbsclient (I suggest you to reinstall the taskworker)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmwm/CRABServer/issues/4583#issuecomment-116506888>.
>

-- 
Stefano Belforte - I.N.F.N.     tel   : +39 040 375-6261 (fax: 375-6258)
Area di Ricerca - Padriciano 99 tel mobile: +39 328 010 7327
34012 TRIESTE TS  - Italy       AIM: stefanobelforte

**Comment by mmascher** [comment]
Hi Stefano,

there is no way to tell which WMCore version has been pulled as the WMCore
code is part of the CRAB3 rpm and it is not installed as an RPM. We
basically bring only the modules we need, not the whole thing.

The tag of WMCore used to build the crabserver RPM is specified in the
specfile of the RPM itself (
https://github.com/cms-sw/cmsdist/blob/comp_gcc481/crabtaskworker.spec) ,
and what we need from WMCore (for each of the CRAB3 components) is
specified here in the setup script of WMCore:
https://github.com/dmwm/WMCore/blob/master/setup_dependencies.py#L119-L140

Long story short make sure you install 3.3.1507.rc1 and you will have the
right version of WMCore!

**Comment by mmascher** [comment]
PS only the taskworker needs this

**Comment by AndresTanasijczuk** [comment]
This commit https://github.com/dmwm/CRABServer/pull/4866 has just been merged into master. It requires WMCore that includes this commit https://github.com/dmwm/WMCore/pull/6056 (currently only in master; no tag).

**Comment by mmascher** [comment]
I have just merged a pull and you will need to update your DB: https://github.com/dmwm/CRABServer/blob/master/etc/updateOracle.sql#L63

**Comment by mmascher** [comment]
A recent WMCore commit is screwing things un in pycurl_manager (see https://github.com/dmwm/WMCore/pull/6142#issuecomment-135361506). Please, don't use the head of WMCore but 1.0.9.pre4, which is the last one we successfully validated.

**Comment by mmascher** [comment]
Hi all,

if you are getting things like:

```
Error contacting the server.
Server answered with: Invalid input parameter
Reason is: The parameter T1_CH_CERN is not in the list of known CMS sites [u'T3_BY_NCPHEP', u'T2_IT_Bari', u'T3_US_Baylor', u'T2_CN_Beijing', u'T3_IT_Bologna', u'T2_UK_SGrid_Bristol', u'T3_US_Brown', u'T2_UK_...
```

keep in mind that because of a recent commit the external configuration parameter `banned-out-destination` is validated against the list of PNN now, so I suggest to you to put something like T1_* in your external conf (wildcards are supported now).

@AndresTanasijczuk , I took the liberty to update https://twiki.cern.ch/twiki/bin/view/CMSPublic/CMSCrabRESTInterface#External_dynamic_configuration with the right values (NB: wildcards will only work with the october release!)

**Comment by mmascher** [comment]
Hi,

I have just merged https://github.com/dmwm/CRABServer/pull/4995 . Make sure you update the "runtim tarballs (htcondor_make_runtime.sh)" if you update TW: I have renamed a classad: https://github.com/mmascher/CRABServer/commit/dfc5155f3f07ec15881e45a07cf114188a56a27a#diff-f7218c126caa1f3f52eaa712fdb9763aL624

**Comment by mmascher** [comment]
Hi all, I have just merged the pull associated to https://github.com/dmwm/CRABServer/issues/4948 . This changes the algorithm used to calculate the checksum of the input sandbox.

Watch out if you are updating to the master branch of client/wmcore/crabserver. The new version of the crabcache is backward compatible, but you cannot use new "client/wmcore" with an old crabcache (the crabcache you use is configured in the external configuration). The risk is that you will get:

```
Server answered with: Input file hashkey mismatch
Reason is: Checksums do not match
```

Also, the new version of the client requires:
https://github.com/dmwm/WMCore/pull/6264
which is not merged in WMCore yet.

Summarizing these are the compatible tags:

Tags:
Client    Crabcache    WMCore
3.3.1510.rc1    3.3.1510.rc2    1.0.9.pre4
3.3.1510.rc2    3.3.1510.rc3    3.3.1510 (from mmascher until a new tag of WMCore is made)

where the crabcache you use is in the external REST configuration (gist). You might be using the preprod or prod crabcache (whcih is the old one!)

**Comment by mmascher** [comment]
Please, change the htcondorPool parameter in your external REST configuration to "cmsgwms-collector-global.cern.ch,cmssrv221.fnal.gov"

That's because vocms099 is going to be retired.
